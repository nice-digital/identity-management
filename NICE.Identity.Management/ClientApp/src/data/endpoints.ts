const API_BASE_URL = process.env.REACT_APP_API_BASE_URL
	? process.env.REACT_APP_API_BASE_URL
	: "#{REACT_APP_API_BASE_URL}";

export const Endpoints = {
	usersList: `${API_BASE_URL}/users`, // get all users
	user: (userId: string | number): string => `${API_BASE_URL}/users/${userId}`,

	servicesList: `${API_BASE_URL}/services`,
	service: (serviceId: string | number): string =>
		`${API_BASE_URL}/services/${serviceId}`,
	
	userRolesByWebsite: (userId: string | number, websiteId: string | number): string =>
		`${API_BASE_URL}/users/${userId}/rolesbywebsite/${websiteId}`,

	verificationEmail: `${API_BASE_URL}/VerificationEmail/VerificationEmail`,
	websitesList: `${API_BASE_URL}/websites`,
};
