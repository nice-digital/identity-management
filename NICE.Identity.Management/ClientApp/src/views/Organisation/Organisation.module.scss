.summaryList {
  display: flex;

  &Label {
    font-weight: bold;
    flex: 1;
  }
  
  &Detail {
    flex: 3;
    display: flex;
    align-self: flex-start;

    > * {
      margin: 0;
    }
  }
}

.userName{
  margin-left:20px;
}
