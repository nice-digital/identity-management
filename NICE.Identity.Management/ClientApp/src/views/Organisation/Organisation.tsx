import React, { Component } from "react";
import { StaticContext } from "react-router";
import { RouteComponentProps, Link } from "react-router-dom";
import { type OrganisationType, type UserType } from "src/models/types";
import { Alert } from "@nice-digital/nds-alert";
import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { PageHeader } from "@nice-digital/nds-page-header";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { Button } from "@nice-digital/nds-button";
import { ErrorMessage } from "src/components/ErrorMessage/ErrorMessage";
import { UserStatus } from "src/components/UserStatus/UserStatus";
import { ToFormattedDateString } from "src/helpers/dateHelpers";
import { fetchData } from "src/helpers/fetchData";
import { isDataError } from "src/helpers/isDataError";
import { Endpoints } from "src/data/endpoints";

import styles from "./Organisation.module.scss";

type TParams = { id: string };

type LocationState = {
	hasBeenEdited: boolean;
};

type OrganisationProps = Record<string, unknown> &
	RouteComponentProps<TParams, StaticContext, LocationState>;

type OrganisationState = {
	organisation: OrganisationType;
	users: Array<UserType>;
	error?: Error;
	isLoading: boolean;
};

export class Organisation extends Component<
	OrganisationProps,
	OrganisationState
> {
	constructor(props: OrganisationProps) {
		super(props);

		this.state = {
			organisation: {} as OrganisationType,
			users: [],
			isLoading: true,
		};

		document.title = "NICE Accounts - Organisation details";
	}

	async componentDidMount(): Promise<void> {
		this.setState({ isLoading: true });

		const organisation = await fetchData(
			Endpoints.organisation(this.props.match.params.id),
		);
		const users = await fetchData(
			Endpoints.usersByOrganisation(this.props.match.params.id),
		);

		if (isDataError(users)) {
			this.setState({ error: users });
		}

		if (isDataError(organisation)) {
			this.setState({ error: organisation });
		}

		this.setState({ organisation, users, isLoading: false });
	}

	render(): JSX.Element {
		const { organisation, users, error, isLoading } = this.state;
		const lastBreadcrumb = isLoading
			? "Loading organisation details"
			: `${organisation.name}`;

		return (
			<>
				<Breadcrumbs>
					<Breadcrumb
						data-qa-sel="breadcrumb-administration-link"
						to="/"
						elementType={Link}
					>
						Administration
					</Breadcrumb>
					<Breadcrumb
						data-qa-sel="breadcrumb-organisation-link"
						to="/organisations"
						elementType={Link}
					>
						Organisations
					</Breadcrumb>
					<Breadcrumb>{lastBreadcrumb}</Breadcrumb>
				</Breadcrumbs>

				{!error ? (
					<>
						<PageHeader
							heading={
								isLoading ? "Organisation details" : `${organisation.name}`
							}
							className="page-header mb--d"
							cta={
								<>
									<Button
										data-qa-sel="edit-organisation-button"
										variant="cta"
										to={`/organisations/${this.props.match.params.id}/edit`}
										elementType={Link}
										disabled={isLoading}
									>
										Edit organisation
									</Button>
									<Button
										data-qa-sel="edit-organisations-users-button"
										to={`/organisations/${this.props.match.params.id}/users`}
										elementType={Link}
										disabled={isLoading}
									>
										Edit users
									</Button>
								</>
							}
						/>

						{isLoading ? (
							<p>Loading...</p>
						) : (
							<>
								{this.props.location?.state?.hasBeenEdited && (
									<Alert
										type="info"
										role="status"
										aria-live="polite"
										data-qa-sel="successful-message-edit-organisation"
									>
										<p>
											The organisation details have been updated successfully.
										</p>
									</Alert>
								)}
								<Grid>
									<GridItem cols={12} md={9} aria-busy={isLoading}>
										<div className={`${styles.summaryList} pv--c`}>
											<span className={styles.summaryListLabel}>
												Date added
											</span>
											<span
												className={styles.summaryListDetail}
												data-qa-sel="dateAdded-organisation"
											>
												{ToFormattedDateString(
													new Date(organisation.dateAdded),
												)}
											</span>
										</div>

										{users.length ? (
											<div className={`${styles.summaryList} pv--c`}>
												<span className={styles.summaryListLabel}>Users</span>
												<span className={styles.summaryListDetail}>
													<ul
														className="list--unstyled"
														data-qa-sel="list-of-organisations"
													>
														{users.map((user) => (
															<li key={user.userId}>
																<UserStatus user={user} />
																<span className={styles.userName}>
																	{`${user.firstName} ${user.lastName}`}
																</span>
															</li>
														))}
													</ul>
												</span>
											</div>
										) : (
											<p>
												There are currently no users assigned to this
												organisation.
											</p>
										)}
									</GridItem>
								</Grid>

								<hr className="mv--b" />
								<h2 className="h3">Permanently delete this organisation</h2>
								<p>
									This organisation will no longer be available, and all
									associated data will be permanently deleted.
								</p>
								<Link
									data-qa-sel="delete-organisation-link"
									to={`/organisations/${organisation.id}/delete`}
									className="pv--c mb--d"
									style={{ display: "inline-block" }}
								>
									Delete organisation
								</Link>
							</>
						)}
					</>
				) : (
					<>
						<PageHeader heading="Error" />
						<ErrorMessage error={error}></ErrorMessage>
					</>
				)}
			</>
		);
	}
}
