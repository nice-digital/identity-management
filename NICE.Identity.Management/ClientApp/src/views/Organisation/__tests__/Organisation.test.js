import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter } from "react-router";
import toJson from "enzyme-to-json";

import { Organisation } from "../Organisation";
import { Endpoints } from "../../../data/endpoints";
import singleOrganisation from "./singleOrganisation.json";
import singleUser from "./singleUser.json";
import { nextTick } from "../../../utils/nextTick";

describe("Organisation", () => {
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};

	const location = {
		state: {}
	};

	const consoleErrorReset = console.error;

	beforeEach(() => {
		fetch.resetMocks();
		console.error = consoleErrorReset;
	});

	it("should show loading message before data has been loaded", () => {
		fetch.mockResponseOnce(JSON.stringify(singleOrganisation));
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		const wrapper = shallow(<Organisation location={location} match={match} />);
		expect(wrapper.find(".OrganisationDateAddedLoadingMsg").text()).toEqual("Loading...")
		expect(wrapper.find(".OrganisationUsersListLoadingMsg").text()).toEqual("Loading...")
	});

	it("should call fetch during componentDidMount", () => {
		fetch.mockResponseOnce(JSON.stringify(singleOrganisation));
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		const wrapper = mount(<MemoryRouter><Organisation location={location} match={match} /></MemoryRouter>);
		const spy = jest.spyOn(wrapper.instance(), "componentDidMount");
		wrapper.instance().componentDidMount();
		wrapper.update();
		expect(spy).toHaveBeenCalled();
		expect(fetch.mock.calls.length).toEqual(1);
		expect(fetch.mock.calls[0][0]).toEqual(Endpoints.organisation(match.params.id));
		spy.mockClear();
	});

	it("should match the snapshot after data has been loaded", async () => {
		fetch.mockResponseOnce(JSON.stringify(singleOrganisation));
		fetch.mockResponseOnce(JSON.stringify(singleUser));
        const wrapper = shallow(<Organisation location={location} match={match} />);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when fetch returns 401 error", async () => {
		console.error = jest.fn();		
		fetch.mockResponseOnce(JSON.stringify({}), { status: 401 });
		fetch.mockResponseOnce(JSON.stringify({}), { status: 401 });
		const wrapper = mount(
			<MemoryRouter>
				<Organisation location={location} match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();		
	});

	it("should show error message when fetch returns 500 error", async () => {
		console.error = jest.fn();
		fetch.mockRejectOnce(new Error("500 Internal Server Error"));
		fetch.mockRejectOnce(new Error("500 Internal Server Error"));
		const wrapper = mount(
			<MemoryRouter>
				<Organisation location={location} match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});
});
