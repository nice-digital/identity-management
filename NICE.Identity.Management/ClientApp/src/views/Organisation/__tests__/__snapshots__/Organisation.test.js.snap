// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organisation should match the snapshot after data has been loaded 1`] = `
<Fragment>
  <Breadcrumbs>
    <Breadcrumb
      data-qa-sel="breadcrumb-organisation-link"
      elementType={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      to="/overview"
    >
      Administration
    </Breadcrumb>
    <Breadcrumb>
      Organisation
    </Breadcrumb>
    <Breadcrumb>
      TEST
    </Breadcrumb>
  </Breadcrumbs>
  <PageHeader
    className="page-header mb--d"
    cta={
      <React.Fragment>
        <Button
          data-qa-sel="edit-organisation-button"
          disabled={false}
          elementType={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          to="/organisations/1/edit"
          variant="cta"
        >
          Edit organisation
        </Button>
        <Button
          data-qa-sel="edit-organisations-users-button"
          disabled={false}
          elementType={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          to="/users-organisations/1/edit"
        >
          Edit users
        </Button>
      </React.Fragment>
    }
    heading="TEST"
  />
  <Grid
    elementType="div"
    equalHeight={false}
    gutter="standard"
    horizontalAlignment="left"
    reverse={false}
    verticalAlignment="top"
  >
    <GridItem
      aria-busy={false}
      cols={12}
      md={9}
    >
      <div
        className="summaryList pv--c"
      >
        <span
          className="summaryListLabel"
        >
          Date added
        </span>
        <span
          className="summaryListDetail"
          data-qa-sel="dateAdded-organisation"
        >
          19 January 2022 17:05
        </span>
      </div>
    </GridItem>
    <GridItem
      aria-busy={false}
      cols={12}
      md={9}
    >
      <p>
        There are currently no users assigned to this organisation.
      </p>
    </GridItem>
  </Grid>
  <hr
    className="mv--b"
  />
  <h2
    className="h3"
  >
    Permanently delete this organisation
  </h2>
  <p>
    This organisation will no longer be available, and all associated data will be permanently deleted.
  </p>
  <Link
    className="pv--c mb--d"
    data-qa-sel="delete-organisation-link"
    style={
      Object {
        "display": "inline-block",
      }
    }
    to="/organisations/1/delete"
  >
    Delete organisation
  </Link>
</Fragment>
`;

exports[`Organisation should show error message when fetch returns 401 error 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            data-qa-sel="breadcrumb-organisation-link"
            href="/overview"
            onClick={[Function]}
          >
            Administration
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Organisation
          </span>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Error
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/overview\\",
        \\"name\\": \\"Administration\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Organisation\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 3,
      \\"item\\": {
        \\"name\\": \\"Error\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
  >
    <h1
      className="page-header__heading"
    >
      Error
    </h1>
  </div>,
  <div
    className="alert alert--error"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>,
]
`;

exports[`Organisation should show error message when fetch returns 500 error 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            data-qa-sel="breadcrumb-organisation-link"
            href="/overview"
            onClick={[Function]}
          >
            Administration
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Organisation
          </span>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Error
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/overview\\",
        \\"name\\": \\"Administration\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Organisation\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 3,
      \\"item\\": {
        \\"name\\": \\"Error\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
  >
    <h1
      className="page-header__heading"
    >
      Error
    </h1>
  </div>,
  <div
    className="alert alert--error"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>,
]
`;
